### ISSUE TITLE
影像向量化模組（雙向量輸出）

### ISSUE THREAD

**Issue by B30780** [B30780/AOI_Defect_LLM_CCPM#35]
﻿**Parent Epic:** #31

name: 影像向量化模組（雙向量輸出）
status: open
created: 2026-01-20T10:40:44Z
updated: 2026-01-20T10:40:44Z
github: [Will be updated when synced to GitHub]
depends_on: [1, 2]
parallel: true
conflicts_with: [5]

# Task: 影像向量化模組（雙向量輸出）

## Description

實作基於 CLIP ViT-B/32 的影像向量化模組，支援雙向量輸出機制（原始影像向量 + 遮罩影像向量）。整合 Qdrant 向量資料庫客戶端，實現高效的向量插入和批次處理。此模組為整個 AOI 缺陷檢索系統的核心特徵提取層。

**核心功能：**
- CLIP ViT-B/32 模型整合與推理
- 原始影像與遮罩影像的雙向量生成
- Qdrant 向量資料庫連接與操作
- 向量正規化與標準化處理
- 批次向量化處理支援
- 向量維度：512-dim

## Acceptance Criteria

- [ ] 成功整合 CLIP ViT-B/32 模型（使用 OpenAI CLIP 或 transformers）
- [ ] 實現雙向量生成：original_vector (512-dim) 和 mask_vector (512-dim)
- [ ] Qdrant 客戶端正常連接並可執行 CRUD 操作
- [ ] 向量正規化到單位向量（L2 norm = 1）
- [ ] 單張影像向量化時間 < 5 秒
- [ ] 批次向量化支援（batch size >= 8）
- [ ] 向量成功儲存到 Qdrant 集合，包含完整 metadata
- [ ] 單元測試覆蓋率 >= 80%
- [ ] 錯誤處理：模型載入失敗、影像格式錯誤、Qdrant 連接失敗

### 架構設計

```python
# src/vectorization/clip_vectorizer.py
import torch
import clip
from PIL import Image
from typing import Tuple, List, Optional
import numpy as np
from qdrant_client import QdrantClient
from qdrant_client.models import PointStruct, VectorParams, Distance

class CLIPVectorizer:
    """
    CLIP-based image vectorizer with dual vector output support.
    Generates embeddings for both original and masked defect images.
    """
    
    def __init__(self, model_name: str = "ViT-B/32", device: str = "cuda"):
        """
        Initialize CLIP model and preprocessing.
        
        Args:
            model_name: CLIP model variant (default: ViT-B/32)
            device: Computation device (cuda/cpu)
        """
        self.device = device
        self.model, self.preprocess = clip.load(model_name, device=device)
        self.model.eval()  # Set to evaluation mode
        
    def vectorize_image(self, image: Image.Image) -> np.ndarray:
        """
        Generate 512-dim embedding for a single image.
        
        Args:
            image: PIL Image object
            
        Returns:
            Normalized 512-dim numpy array
        """
        with torch.no_grad():
            image_input = self.preprocess(image).unsqueeze(0).to(self.device)
            embedding = self.model.encode_image(image_input)
            embedding = embedding / embedding.norm(dim=-1, keepdim=True)  # L2 normalize
            return embedding.cpu().numpy().flatten()
    
    def vectorize_dual(self, 
                       original_image: Image.Image, 
                       mask_image: Image.Image) -> Tuple[np.ndarray, np.ndarray]:
        """
        Generate dual vectors for original and masked images.
        
        Args:
            original_image: Original defect image
            mask_image: Masked defect region image
            
        Returns:
            Tuple of (original_vector, mask_vector), both 512-dim
        """
        original_vector = self.vectorize_image(original_image)
        mask_vector = self.vectorize_image(mask_image)
        return original_vector, mask_vector
    
    def batch_vectorize(self, images: List[Image.Image], 
                       batch_size: int = 8) -> np.ndarray:
        """
        Batch vectorization for efficiency.
        
        Args:
            images: List of PIL Image objects
            batch_size: Batch processing size
            
        Returns:
            Array of shape (N, 512)
        """
        embeddings = []
        for i in range(0, len(images), batch_size):
            batch = images[i:i+batch_size]
            with torch.no_grad():
                batch_input = torch.stack([
                    self.preprocess(img) for img in batch
                ]).to(self.device)
                batch_embeddings = self.model.encode_image(batch_input)
                batch_embeddings = batch_embeddings / batch_embeddings.norm(
                    dim=-1, keepdim=True
                )
                embeddings.append(batch_embeddings.cpu().numpy())
        
        return np.vstack(embeddings)
```

### Qdrant 整合

```python
# src/vectorization/qdrant_manager.py
from qdrant_client import QdrantClient
from qdrant_client.models import (
    PointStruct, VectorParams, Distance, 
    NamedVector, PayloadSchemaType
)
from typing import Dict, Any, List
import uuid

class QdrantVectorManager:
    """
    Manager for Qdrant vector database operations.
    Handles dual vector storage (original + mask).
    """
    
    def __init__(self, host: str = "localhost", port: int = 6333):
        self.client = QdrantClient(host=host, port=port)
        
    def create_collection(self, collection_name: str):
        """
        Create Qdrant collection with dual vector support.
        """
        self.client.recreate_collection(
            collection_name=collection_name,
            vectors_config={
                "original": VectorParams(
                    size=512,
                    distance=Distance.COSINE
                ),
                "mask": VectorParams(
                    size=512,
                    distance=Distance.COSINE
                )
            }
        )
        
    def insert_dual_vector(self,
                          collection_name: str,
                          original_vector: np.ndarray,
                          mask_vector: np.ndarray,
                          metadata: Dict[str, Any]) -> str:
        """
        Insert dual vectors with metadata.
        
        Args:
            collection_name: Target collection
            original_vector: 512-dim original image vector
            mask_vector: 512-dim mask image vector
            metadata: Additional metadata (file_path, defect_type, etc.)
            
        Returns:
            UUID of inserted point
        """
        point_id = str(uuid.uuid4())
        
        self.client.upsert(
            collection_name=collection_name,
            points=[
                PointStruct(
                    id=point_id,
                    vector={
                        "original": original_vector.tolist(),
                        "mask": mask_vector.tolist()
                    },
                    payload=metadata
                )
            ]
        )
        
        return point_id
    
    def batch_insert(self,
                    collection_name: str,
                    vectors_data: List[Dict[str, Any]],
                    batch_size: int = 100):
        """
        Batch insert for efficiency.
        
        Args:
            collection_name: Target collection
            vectors_data: List of dicts with keys: original_vector, mask_vector, metadata
            batch_size: Batch size for insertion
        """
        for i in range(0, len(vectors_data), batch_size):
            batch = vectors_data[i:i+batch_size]
            points = [
                PointStruct(
                    id=str(uuid.uuid4()),
                    vector={
                        "original": data["original_vector"].tolist(),
                        "mask": data["mask_vector"].tolist()
                    },
                    payload=data["metadata"]
                )
                for data in batch
            ]
            self.client.upsert(collection_name=collection_name, points=points)
```

### 向量化流程整合

```python
# src/vectorization/vectorization_pipeline.py
from pathlib import Path
from PIL import Image
from typing import Optional
import logging

class VectorizationPipeline:
    """
    End-to-end vectorization pipeline integrating CLIP and Qdrant.
    """
    
    def __init__(self, 
                 qdrant_host: str = "localhost",
                 qdrant_port: int = 6333,
                 collection_name: str = "aoi_defects",
                 device: str = "cuda"):
        self.vectorizer = CLIPVectorizer(device=device)
        self.qdrant = QdrantVectorManager(host=qdrant_host, port=qdrant_port)
        self.collection_name = collection_name
        self.logger = logging.getLogger(__name__)
        
    def initialize_collection(self):
        """Initialize Qdrant collection if not exists."""
        self.qdrant.create_collection(self.collection_name)
        self.logger.info(f"Collection '{self.collection_name}' initialized")
    
    def process_image_pair(self,
                          original_path: Path,
                          mask_path: Path,
                          metadata: Optional[Dict] = None) -> str:
        """
        Process a pair of images and store in Qdrant.
        
        Args:
            original_path: Path to original image
            mask_path: Path to mask image
            metadata: Additional metadata to store
            
        Returns:
            UUID of inserted point
        """
        try:
            # Load images
            original_img = Image.open(original_path).convert('RGB')
            mask_img = Image.open(mask_path).convert('RGB')
            
            # Generate dual vectors
            original_vec, mask_vec = self.vectorizer.vectorize_dual(
                original_img, mask_img
            )
            
            # Prepare metadata
            full_metadata = {
                "original_path": str(original_path),
                "mask_path": str(mask_path),
                "image_size": original_img.size,
                **(metadata or {})
            }
            
            # Insert into Qdrant
            point_id = self.qdrant.insert_dual_vector(
                self.collection_name,
                original_vec,
                mask_vec,
                full_metadata
            )
            
            self.logger.info(f"Vectorized and stored: {original_path.name}")
            return point_id
            
        except Exception as e:
            self.logger.error(f"Error processing {original_path}: {str(e)}")
            raise
```

### 效能優化考量

1. **批次處理**：使用 batch_vectorize 減少 GPU 傳輸次數
2. **向量正規化**：使用 L2 正規化確保餘弦相似度計算準確
3. **記憶體管理**：使用 torch.no_grad() 避免梯度計算
4. **異步處理**：大量影像可使用多執行緒或 async IO

### 測試檔案

```python
# tests/test_vectorization.py
import pytest
import numpy as np
from PIL import Image
from src.vectorization.clip_vectorizer import CLIPVectorizer

def test_single_image_vectorization():
    """Test single image vectorization output shape and normalization."""
    vectorizer = CLIPVectorizer(device="cpu")
    test_image = Image.new('RGB', (224, 224), color='red')
    
    vector = vectorizer.vectorize_image(test_image)
    
    assert vector.shape == (512,), "Vector dimension should be 512"
    assert np.isclose(np.linalg.norm(vector), 1.0, atol=1e-5), "Vector should be normalized"

def test_dual_vectorization():
    """Test dual vector generation."""
    vectorizer = CLIPVectorizer(device="cpu")
    original = Image.new('RGB', (224, 224), color='blue')
    mask = Image.new('RGB', (224, 224), color='green')
    
    orig_vec, mask_vec = vectorizer.vectorize_dual(original, mask)
    
    assert orig_vec.shape == (512,)
    assert mask_vec.shape == (512,)
    assert not np.allclose(orig_vec, mask_vec), "Vectors should be different"
```

### 相關檔案

- `src/vectorization/clip_vectorizer.py` - CLIP 向量化器
- `src/vectorization/qdrant_manager.py` - Qdrant 管理器
- `src/vectorization/vectorization_pipeline.py` - 整合流程
- `tests/test_vectorization.py` - 單元測試
- `config/vectorization_config.yaml` - 配置檔案

### 前置條件
- [1] 資料集前處理模組完成（提供影像資料）
- [2] 資料管理與配置模組完成（提供資料路徑與配置）

### 技術依賴
- Python >= 3.8
- PyTorch >= 1.12.0
- OpenAI CLIP (`pip install git+https://github.com/openai/CLIP.git`)
- Qdrant Client (`pip install qdrant-client`)
- Pillow >= 9.0.0
- NumPy >= 1.3.0

### 外部服務
- Qdrant 向量資料庫（Docker 部署或雲端服務）
- CUDA 支援的 GPU（建議用於加速，CPU 也可運行但較慢）

## Effort Estimate

- **Size**: M (Medium)
- **Hours**: 7 hours (3 days)
  - CLIP 整合與測試：8 hours
  - Qdrant 客戶端開發：6 hours
  - 雙向量機制實作：4 hours
  - 批次處理與優化：4 hours
  - 測試與文檔：2 hours
- **Parallel**: true（可與 Task 5 並行開發，但需注意 GPU 資源分配）
- **Conflicts with**: [5]（兩者都需要 GPU 資源，建議錯開測試時間）

## Definition of Done

- [ ] 程式碼實作完成並遵循 PEP 8 規範
- [ ] CLIP 模型成功載入並可生成 512-dim 向量
- [ ] Qdrant 集合建立成功，支援雙向量儲存
- [ ] 單張影像向量化效能 < 5 秒（GPU）
- [ ] 批次向量化功能正常運作（batch_size >= 8）
- [ ] 單元測試撰寫完成，覆蓋率 >= 80%
- [ ] 整合測試通過：端到端向量化流程
- [ ] 效能測試通過：1000 張影像批次處理 < 30 分鐘
- [ ] API 文檔完成（docstrings + README）
- [ ] 程式碼審查通過
- [ ] 與 Task 2 的配置模組成功整合
